rmf_scheduler_ros2/src/Publisher.cpp

```diff
@@ -0,0 +1,82 @@
+/*
+ * Copyright (C) 2022 Open Source Robotics Foundation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+#include <rclcpp/serialization.hpp>
+
+#include <rmf_scheduler_msgs/msg/serialized_message.hpp>
+
+#include "Publisher.hpp"
+
+namespace rmf::scheduler {
+
+Publisher::Publisher(rclcpp::Node* node)
+: node(node)
+{
+}
+
+void Publisher::publish(const rmf_scheduler_msgs::msg::Payload& payload)
+{
+  switch (payload.type)
+  {
+    case rmf_scheduler_msgs::msg::Payload::PAYLOAD_TYPE_SERIALIZED_MESSAGE:
+      return this->_publish_serialized_message(payload.data);
+    default:
+      RCLCPP_ERROR(
+        this->node->get_logger(), "Unsupported payload type: %d", payload.type);
+  }
+}
+
+void Publisher::_publish_serialized_message(const PayloadData& data)
+{
+  // This function can be confusing due to the multiple layers of serialization.
+  //
+  // First, the data from the param is the serialized rmf_scheduler_msgs/SerializedMessage.
+  // We need to deserialize this to get the concrete rmf_scheduler_msgs/SerializedMessage.
+  //
+  // Within rmf_scheduler_msgs/SerializedMessage, there is another blob which contains the
+  // serialized form of the message to publish. This serialized message will be published
+  // with the generic publisher.
+
+  using Serializer =
+    rclcpp::Serialization<rmf_scheduler_msgs::msg::SerializedMessage>;
+  static Serializer serializer;
+
+  rclcpp::SerializedMessage payload_sermsg{data.size()};
+  auto& rcl_payload =
+    payload_sermsg.get_rcl_serialized_message();
+  std::copy(&data.front(), &data.back(), rcl_payload.buffer);
+  rcl_payload.buffer_length = data.size();
+
+  // deserialize the serialized rmf_scheduler_msgs/SerializedMessage
+  rmf_scheduler_msgs::msg::SerializedMessage rmf_msg;
+  auto asd = payload_sermsg.size();
+  serializer.deserialize_message(&payload_sermsg, &rmf_msg);
+
+  rclcpp::SerializedMessage inner_sermsg{rmf_msg.data.size()};
+  auto& rcl_inner =
+    inner_sermsg.get_rcl_serialized_message();
+  std::copy(rmf_msg.data.data(),
+    rmf_msg.data.data() + rmf_msg.data.size(), rcl_inner.buffer);
+  rcl_inner.buffer_length = rmf_msg.data.size();
+
+  auto pub = this->node->create_generic_publisher(rmf_msg.topic_name,
+      rmf_msg.message_type,
+      rclcpp::SystemDefaultsQoS{});
+  pub->publish(inner_sermsg);
```

Comment:
Gave it a quick spin and looking good so far!
I have a mix of feedback and question on reliability since I see that the approach is to create a publisher, publish a message then destroy it immediately.

I observed that "sometimes" some of my subscribers don't get the messages even though they are subscribing to the topic being published. _I believe_ this might be a discovery issue, since the publisher creation and subscription immediately follow one another it might happen that all the subscribers were not discovered.
I tested this case by running a few subscribers in parallel (i.e. a few applications, a `ros2 topic echo`) and adding a trigger to be ran as soon as possible (in the past), together with a trigger to run a bit in the future. Sometimes one of my applications (really only the `ros2 topic echo`) misses the first message, but seems to always work for the second message (maybe due to some caching happening at the ros2 layer for subscribers?).

Adding a very hacky `usleep(1e6)` between publisher creation and message publishing worked but that is most likely not a great way to deal with this

Reply:
Hmm I'm not sure how to go about addressing this. The node is created early and stays alive throughout so I thought this won't be an issue, at least it works on the e2e test :thinking:. Can you try the e2e tests and see if it works on your pc? Even better, it would be really helpful if we can reproduce it in the tests.

I'm thinking a possible solution is to wait for subscribe count > 0 before publishing, but that may only work when considering only 1 subscriber, when there are more subscribers we don't know how many subscribers to be discovered before publishing.

We can keep the publisher alive but I'm not sure if it will work because afaik discovery is based on nodes and not publishers. Even if it works, it wouldn't be a good solution as we will end up with a publisher for each trigger/schedule, possibly resulting in hundreds or maybe even thousands of publishers.

Reply:
The E2E tests ran through `colcon test --packages-select rmf_scheduler_ros2_client` all fail on my machine and I'm not quite sure why:

```
<?xml version="1.0" encoding="UTF-8"?>
<testsuites>
  <testsuite name="e2e_TEST" errors="0" failures="5" tests="5" hostname="tbd" time="6.547" timestamp="2022-04-21T09:33:40Z">
    <testcase classname="e2e_TEST.global" name="create trigger" time="0.000" status="run">
      <failure type="FAIL">
FAILED:
Timed out waiting for new trigger to be executed
at /home/luca/ws_avcs/src/rmf/rmf_ros2/rmf_scheduler_ros2_client/test/e2e_TEST.cpp:139
      </failure>
    </testcase>
    <testcase classname="e2e_TEST.global" name="cancel_schedule" time="0.000" status="run">
      <failure type="FAIL">
FAILED:
Timed out waiting for response
at /home/luca/ws_avcs/src/rmf/rmf_ros2/rmf_scheduler_ros2_client/test/e2e_TEST.cpp:79
      </failure>
    </testcase>
    <testcase classname="e2e_TEST.global" name="list schedules" time="0.000" status="run">
      <failure type="FAIL">
FAILED:
Timed out waiting for response
at /home/luca/ws_avcs/src/rmf/rmf_ros2/rmf_scheduler_ros2_client/test/e2e_TEST.cpp:79
      </failure>
    </testcase>
    <testcase classname="e2e_TEST.global" name="list schedule states" time="0.000" status="run">
      <failure type="FAIL">
FAILED:
Timed out waiting for response
at /home/luca/ws_avcs/src/rmf/rmf_ros2/rmf_scheduler_ros2_client/test/e2e_TEST.cpp:79
      </failure>
    </testcase>
    <testcase classname="e2e_TEST.global" name="cancel all" time="0.000" status="run">
      <failure type="FAIL">
FAILED:
Timed out waiting for response
at /home/luca/ws_avcs/src/rmf/rmf_ros2/rmf_scheduler_ros2_client/test/e2e_TEST.cpp:79
      </failure>
    </testcase>
    <system-out/>
    <system-err/>
  </testsuite>
</testsuites>
```
I even get `uncrustify` errors:

```
luca@luxury:~/ws_avcs$ colcon test-result
build/rmf_scheduler_ros2_client/Testing/20220421-0935/Test.xml: 2 tests, 0 errors, 2 failures, 0 skipped
build/rmf_scheduler_ros2_client/test_results/rmf_scheduler_ros2_client/e2e_TEST.catch2.xml: 5 tests, 0 errors, 5 failures, 0 skipped
build/rmf_scheduler_ros2_client/test_results/rmf_scheduler_ros2_client/uncrustify.xunit.xml: 2 tests, 0 errors, 1 failure, 0 skipped

Summary: 65100 tests, 0 errors, 8 failures, 0 skipped
```

I also find it a bit suspicious that CI is green, I can't figure out how building can pass if the dependency PR under rmf_internal_msgs is not merged, shouldn't compilation fail because it fails to find the message headers?

Reply:
> I'm thinking a possible solution is to wait for subscribe count > 0 before publishing, but that may only work when considering only 1 subscriber, when there are more subscribers we don't know how many subscribers to be discovered before publishing.

I think this is an OK solution, I am a bit worried of the edge case of having no subscribers and what would happen, I guess an "async wait with a timeout for num_subscribers > 0" should cover most cases?

Reply:
I wouldn't depend on this if possible, this solution doesn't really work because we can't know how many subscribers there will be, if we published as soon as there is 1 subscriber, the 2nd subscriber will not receive the message, leading to the same problem again.

A better solution is to have subscribers not rely on one off messages during its startup sequence or fix the launch system to ensure targets are properly started in order.